pandas/_libs/reshape.pyx

```diff
@@ -1,15 +1,95 @@
 # -*- coding: utf-8 -*-
 
-cimport cython
-from cython cimport Py_ssize_t
+import cython
+from cython import Py_ssize_t
 
-import numpy as np
-from numpy cimport (ndarray,
-                    int8_t, int16_t, int32_t, int64_t, uint8_t, uint16_t,
+from numpy cimport (int8_t, int16_t, int32_t, int64_t, uint8_t, uint16_t,
                     uint32_t, uint64_t, float32_t, float64_t)
 
 
-cdef double NaN = <double> np.NaN
-cdef double nan = NaN
+ctypedef fused reshape_t:
+    uint8_t
+    uint16_t
+    uint32_t
+    uint64_t
+    int8_t
+    int16_t
+    int32_t
+    int64_t
+    float32_t
+    float64_t
+    object
 
-include "reshape_helper.pxi"
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def unstack(reshape_t[:, :] values, uint8_t[:] mask,
+            Py_ssize_t stride, Py_ssize_t length, Py_ssize_t width,
+            reshape_t[:, :] new_values, uint8_t[:, :] new_mask):
+    """
+    transform long sorted_values to wide new_values
+
+    Parameters
+    ----------
+    values : typed ndarray
+    mask : boolean ndarray
+    stride : int
+    length : int
+    width : int
+    new_values : typed ndarray
+        result array
+    new_mask : boolean ndarray
+        result mask
+    """
+    cdef:
+        Py_ssize_t i, j, w, nulls, s, offset
+
+    if reshape_t is not object:
```

Comment:
what if we make the inner a function instead?

Reply:
No dice; we'd need a "nogil" version and a not-nogil version.

Reply:
I think like long-term cython may implement a feature discussed in #22452 that may let us clear this up further.  For now this is much more readable than the existing tempita implementation.